### PR TITLE
Add Skygear Chat UIKit Android & iOS Guide

### PR DESCRIPTION
**Pull request by howawong**
connects #165

### CODE REVIEW

**Comment by carmenlau** [comment]
In both iOS and Android installation instructions should mention the required permission setup

**Comment by howawong** [comment]
@carmenlau, in Android, `chat` and `chat_ui` `AndroidManifest.xml` include the permission already.

**Comment by howawong** [comment]
@carmenlau, I checked with @ben181231, it does not take extra permission except push notification.

File: chat/uikit/android.md [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+import io.skygear.plugins.chat.ui.*
+```
+
+
+## Creating a conversation UI
+The following code block creates a `ConversationActivity` of a conversation `c`.
+
+```java
+Intent intent = new Intent(this, ConversationActivity.class);
+intent.putExtra(ConversationActivity.ConversationIntentKey, c.toJson().toString()); // c is an instance of an io.skygear.plugins.chat.Conversation
```
Comment (carmenlau) [inline chat/uikit/android.md]: Please put the comment to another line, it is difficult to read :P

File: chat/uikit/android.md [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+startActivity(intent);
+```
+
+```kotlin
+val intent = Intent(this, ConversationActivity::class.java)
+intent.putExtra(ConversationActivity.ConversationIntentKey, c.toJson().toString())
```
Comment (carmenlau) [inline chat/uikit/android.md:67]: Please add the comment to the kotlin version too

File: chat/uikit/android.md [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+|`MessageSentListenerIntentKey`|MessageSentListener|
+|`MessageFetchListenerIntentKey`|MessageFetchListener|
+|`ConnectionListenerIntentKey`|ConnectionListener|
+
+
+## Adapter
+UIKit provides `ConversationViewAdapter` to enable remote HTTP background image in `ConversationActivity`.
+
+```java
+ConversationViewAdapter adapter = new ConversationViewAdapter() {
```
Comment (carmenlau) [inline chat/uikit/android.md:183]: Missing example for how to set `ConversationViewAdapter` to Conversation Activity?

File: chat/uikit/ios.md [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
+```swift
+optional func conversationViewController(
+    _ controller: SKYChatConversationViewController,
+    dateStringAt indexPath: IndexPath) -> NSAttributedString
+```
+
+- To enable or disable date
+
+```obj-c
+(bool) conversationViewController:(SKYChatConversationViewController *)controller shouldShowDateAt:(NSIndexPath *)indexPath
```
Comment (carmenlau) [inline chat/uikit/ios.md]: Missing `-` before `(bool)`

File: chat/uikit/ios.md [hunk 5]
```diff
@@ -0,0 +1,386 @@
...
+
+
+## Delegates
+You may also implement functions in `SKYChatConversationViewControllerDelegate`
+
+### Message Date
+
+- To customize date format
+```obj-c
+- (NSAttributedString *) conversationViewController:(SKYChatConversationViewController *)controller dateStringAt:(NSIndexPath *)indexPath
```
Comment (carmenlau) [inline chat/uikit/ios.md:201]: Please put implementation example for those delegate functions which are not just return bool. That would be easier to understand :)

File: chat/uikit/ios.md [hunk 6]
```diff
@@ -0,0 +1,386 @@
...
+
+
+## SKYChatUIModelCustomization
+You can change `userNameField` and `userAvatarField` via `SKYChatUIModelCustomization.default()`.
+
+
+```obj-c
+[[[SKYChatUIModelCustomization default]
+    updateWithUserNameField:@"name"]
+    updateWithUserAvatarField:@"profile_pic"];
```
Comment (carmenlau) [inline chat/uikit/ios.md]: I think it is not common to chain the call this way in objective-c? Would be better to call the update separately like this?
```
[[SKYChatUIModelCustomization default] updateWithUserNameField:@"name"];
[[SKYChatUIModelCustomization default] updateWithUserAvatarField:@"profile_pic"];
```

File: chat/uikit/ios.md [hunk 7]
```diff
@@ -0,0 +1,386 @@
...
+
+Please follow the steps at [Quick Start](https://docs.skygear.io/guides/chat/quick-start/ios/).
+
+### Step 2: Add Skygear Chat UIKit to your project
+
+
+1. add `pod 'SKYKitChat/UI', '~> 1.1'` in `Podfile`.
+
+```bash
+pod 'SKYKitChat/UI', '~> 1.1'
```
Comment (carmenlau) [inline chat/uikit/ios.md]: Just afraid we will forget to update this version, we should be a newer version that include the ui kit update. Should be 1.2 or 1.4? We can discuss.

**Review by carmenlau — CHANGES_REQUESTED** [review]
The content is good:)! But the section organization makes me a bit difficult to find the information i want.

Currently ios has following sections:
```
About UIKit
Prerequisite
Creating a conversation UI
Customizations
UICustomization
SKYChatUIModelCustomization
Delegates
```
And android:
```
About UIKit
Prerequisite
Creating a conversation UI
Customizations
Attribute Customization
Event Hooks
Adapter
Avatar Adapter
```

For example, ui customization and event hooks are quite different area. Maybe we should separate them clearly? 
Feel free to discuss how to make better organization for the sections.

File: chat/uikit/ios.md (hunk 7 above)
Comment (howawong) [reply chat/uikit/ios.md]: @carmenlau  what is the latest stable release? Is it 1.4 ? If yes, then we just put 1.4.

Comment (howawong) [reply chat/uikit/ios.md]: @rickmak  any thought on versioning?

File: chat/uikit/ios.md (hunk 5 above)
Comment (howawong) [reply chat/uikit/ios.md:201]: updated.

File: chat/uikit/android.md (hunk 3 above)
Comment (howawong) [reply chat/uikit/android.md:183]: updated.

File: chat/uikit/android.md (hunk 2 above)
Comment (howawong) [reply chat/uikit/android.md:67]: updated.

**Comment by howawong** [comment]
@carmenlau  updated version in `Podfile` to 1.4

File: chat/uikit/ios.md [hunk 8]
```diff
@@ -0,0 +1,408 @@
...
+
+The following properties are supported.
+### Avatar
+|Attribute|`Description`|Type|
+|---------|-----------|------|
+|`avatarBackgroundColor`|Background color of the avatar|Color|
+|`avatarTextColor`|Text color of the avatar|Color|
+|`userAvatarType`|Avatar Type, either showing initial of the username field or image|`SKYChatConversationViewUserAvatarType`, "initial" or "image"|
+|`avatarHiddenForIncomingMessages`|Hide incoming message avatar if true|Boolean, default False|
+|`avatarHiddenForOutgoingMessages`|Hide outgoing message avatar if true|Boolean, default True|
```
Comment (carmenlau) [inline chat/uikit/ios.md]: Check the code the `avatarHiddenForOutgoingMessages` default is False

File: chat/uikit/ios.md [hunk 9]
```diff
@@ -0,0 +1,408 @@
...
+
+### Status Text & Date Time Format
+|Attribute|`Description`|Type|
+|---------|-----------|------|
+|`messageStatusDelivering`|Delivering text|String, default "Delivering"|
+|`messageStatusDelivered`|Delivered text|String, default "Delivered"|
+|`messageStatusSomeRead`|Some read text|String, default "Some Read"|
+|`messageStatusAllRead`|All read text|String, default "All Read"|
+|`messageSentFailed`|Failed text|String, default "Failed"|
+|`messageDateFormatter`|Message delivery time date formatter|`DateFormatter`|
```
Comment (carmenlau) [inline chat/uikit/ios.md]: Text customization is under `textCustomization` in  not `UICustomization`

File: chat/uikit/ios.md [hunk 10]
```diff
@@ -0,0 +1,408 @@
...
+SKYChatConversationView.UICustomization().avatarHiddenForIncomingMessages = false
+```
+
+The following properties are supported.
+### Avatar
+|Attribute|`Description`|Type|
+|---------|-----------|------|
+|`avatarBackgroundColor`|Background color of the avatar|Color|
+|`avatarTextColor`|Text color of the avatar|Color|
+|`userAvatarType`|Avatar Type, either showing initial of the username field or image|`SKYChatConversationViewUserAvatarType`, "initial" or "image"|
```
Comment (carmenlau) [inline chat/uikit/ios.md]: No `userAvatarType`, it should be `avatarType`.

Please also add how to select the avatar field in this case

Comment (carmenlau) [reply chat/uikit/ios.md]: It's still showing user initial with following config
 
```
 [[SKYChatUIModelCustomization default] updateWithUserAvatarField:@"profile_pic"];
 [SKYChatConversationView UICustomization].avatarType = SKYChatConversationViewUserAvatarTypeImage;
```

File: chat/uikit/ios.md [hunk 11]
```diff
@@ -0,0 +1,408 @@
...
+
+
+##### `SKYChatConversationViewController`''s properties
+
+You may also customize `SKYChatConversationViewController`s properties before it is displayed.
+
+```obj-c
+SKYChatConversationViewController* conversationViewController = [[SKYChatConversationViewController alloc] init];
+conversationViewController.conversation = c;
+conversationViewController.incomingMessageBubbleColor = [UIColor redColor];
```
Comment (carmenlau) [inline chat/uikit/ios.md]: We are not using this way to customize `incomingMessageBubbleColor` this way. We used customization class and delegate

File: chat/uikit/ios.md [hunk 12]
```diff
@@ -0,0 +1,408 @@
...
+
+
+##### Feature On/Off
+|Attribute|`Description`|Type|
+|---------|-----------|------|
+|`shouldShowCameraButton`|Show camera button if true|Bool, default: True|
+|`shouldShowVoiceMessageButton`|Show voice message button if true|Bool, default: True|
+|`shouldShowTypingIndicator`|Show typing indicators if true|Bool, default: True|
+|`shouldShowMessageStatus`|Show message status if true|Bool, default: true|
+|`shouldShowAccessoryButton`|Show accessory button if true|Bool, default: true|
```
Comment (carmenlau) [inline chat/uikit/ios.md]: Those name are not correct, from the code they are `cameraButtonShouldShow`

File: chat/uikit/ios.md [hunk 13]
```diff
@@ -0,0 +1,408 @@
...
+|`incomingMessageBubbleColor`|Incoming messages background color|Color|
+|`outgoingMessageBubbleColor`|Outgoing messages background color|Color|
+
+
+##### Feature On/Off
+|Attribute|`Description`|Type|
+|---------|-----------|------|
+|`shouldShowCameraButton`|Show camera button if true|Bool, default: True|
+|`shouldShowVoiceMessageButton`|Show voice message button if true|Bool, default: True|
+|`shouldShowTypingIndicator`|Show typing indicators if true|Bool, default: True|
```
Comment (carmenlau) [inline chat/uikit/ios.md]: Please use `true` instead of `True`

File: chat/uikit/ios.md (hunk 12 above)
Comment (carmenlau) [reply chat/uikit/ios.md]: We don't have accessoryButton setting in UICustomization

Comment (carmenlau) [reply chat/uikit/ios.md]: I think we should add `accessoryButtonShouldShow` for customization class to keep the api consistent. User we use both customization class and delegate for this setting. Let me open issue in iOS that this. Please keep it in the document.

File: chat/uikit/ios.md (hunk 9 above)
Comment (carmenlau) [reply chat/uikit/ios.md]: Missing send button

Comment (carmenlau) [reply chat/uikit/ios.md]: `messageDateFormatter` is not in this section, it is under UICustomization

File: chat/uikit/ios.md [hunk 14]
```diff
@@ -0,0 +1,408 @@
...
+
+```obj-c
+- (NSAttributedString *) conversationViewController:(SKYChatConversationViewController *)controller dateStringAt:(NSIndexPath *)indexPath {
+    SKYMessage* msg = [self messageList] messageAt: indexPath.row];
+    NSDateFormatter * formatter = [[NSDateFormatter alloc] init];
+    [formatter setDateFormat: @"yyyy-MM-dd"];
+    NSString *dateString = [formatter stringFromDate: [msg creationDate]];
+    NSMutableAttributedString *attributedString = [[NSMutableAttributedString alloc] initWithString:dateString];
+    [attributedString setTextColor:[UIColor blueColor]];
+    return attributedString;
```
Comment (carmenlau) [inline chat/uikit/ios.md]:
```
    SKYMessage* msg = [[controller messageList] messageAt: indexPath.row];
    NSDateFormatter * formatter = [[NSDateFormatter alloc] init];
    [formatter setDateFormat: @"yyyy-MM-dd"];
    NSString *dateString = [formatter stringFromDate: [msg creationDate]];

    NSDictionary *attrs = @{ NSForegroundColorAttributeName : [UIColor blueColor] };
    NSAttributedString *attrStr = [[NSAttributedString alloc] initWithString:dateString attributes:attrs];
    return attrStr;
```

**Review by carmenlau — CHANGES_REQUESTED** [review]

**Comment by carmenlau** [comment]
We also missed the code to teach user how to set the conversation view background 🙈

**Comment by carmenlau** [comment]
The background image code for your reference:

Customization class
```
    [SKYChatConversationView UICustomization].backgroundColor = [UIColor lightGrayColor];
    [SKYChatConversationView UICustomization].backgroundImage = [UIImage imageNamed:<imageName>];
    [SKYChatConversationView UICustomization].backgroundImageURL = [NSURL URLWithString:<imageURL>];
```

Delegate
```
- (UIColor*) backgroundColorForConversationViewController:(SKYChatConversationViewController *)controller {
    return [UIColor lightGrayColor];
}

- (UIImage*) backgroundImageForConversationViewController:(SKYChatConversationViewController *)controller {
    return [UIImage imageNamed:<imageName>];
}

- (NSURL*) backgroundImageURLForConversationViewController:(SKYChatConversationViewController *)controller {
    return [NSURL URLWithString:<imageURL>];
}
```

**Comment by carmenlau** [comment]
@howawong I found we only includes some of the event hooks here, we have more events hooks actually. Also do you think it would be better to open another page for UIKit event hooks? If you agree, let's remove the event hooks from this pr and we will focuses on the ui customization in this PR first. And another pr and issue for the event hooks?

btw, the participants hooks is not updated, the function should be `didFetchParticipants` and receive list of `SKYParticipant` instead of records

https://github.com/SkygearIO/chat-SDK-iOS/blob/02aa80cca37b61708bb7625640021f98e9f4a55f/SKYKitChat/Classes/UI/ConversationView/SKYChatConversationViewController.swift#L123

**Comment by carmenlau** [comment]
@ben181231 Would like to have your help to check the android part

**Comment by howawong** [comment]
Merged to `next` branch via `ce2a33f0c3f2fb8abfd4c1df8ee296f4f105b779`. Please create an issue if needed.

**Pull request closed without merging**